### PR TITLE
feat!: use struct variants for blackbox function calls

### PR DESCRIPTION
**Pull request by sirasistant**
# Related issue(s)

# Description

## Summary of changes
Define different types for the variants of BB function call
## Test additions / changes

# Checklist

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [ ] I have reviewed the changes on GitHub, line by line.
- [ ] I have ensured all changes are covered in the description.

# Additional context

(If applicable.)

### CODE REVIEW

File: acvm/src/lib.rs [hunk 1]
```diff
@@ -99,37 +100,46 @@ pub trait PartialWitnessGenerator {
...
+        public_key_x: &FunctionInput,
+        public_key_y: &FunctionInput,
+        signature: &[FunctionInput],
+        message: &[FunctionInput],
+        output: &Witness,
     ) -> Result<pwg::OpcodeResolution, OpcodeResolutionError>;
     fn pedersen(
         &self,
         initial_witness: &mut BTreeMap<Witness, FieldElement>,
+        hash_index: u32,
```
Comment (kevaundray) [inline acvm/src/lib.rs:120]: We would want to put this in as a separate change since it is unrelated to changing ACIR to use struct variants

File: acvm/src/pwg/signature/ecdsa.rs [hunk 2]
```diff
@@ -3,45 +3,42 @@ use std::collections::BTreeMap;
 
 use crate::{pwg::witness_to_value, pwg::OpcodeResolution, OpcodeResolutionError};
 
-pub fn secp256k1_prehashed(
+fn to_u8_vec(
```
Comment (TomAFrench) [inline acvm/src/pwg/signature/ecdsa.rs:6]: `initial_witness` can be immutable here.

File: acvm/src/pwg/signature/ecdsa.rs [hunk 3]
```diff
@@ -3,45 +3,42 @@ use std::collections::BTreeMap;
...
-
-        let y_i = witness_to_value(initial_witness, _y_i.witness)?;
-        *pky = *y_i.to_be_bytes().last().unwrap();
+    value: &[FunctionInput],
+) -> Result<Vec<u8>, OpcodeResolutionError> {
+    let mut result = Vec::new();
+    for input in value {
+        let w_value = witness_to_value(initial_witness, input.witness)?.to_be_bytes();
+        let byte = w_value.last().unwrap();
+        result.push(*byte);
```
Comment (TomAFrench) [inline acvm/src/pwg/signature/ecdsa.rs:14]:
```suggestion
    inputs: &[FunctionInput],
) -> Result<Vec<u8>, OpcodeResolutionError> {
    let mut result = Vec::with_capacity(inputs.len());
    for input in inputs {
        let witness_value = witness_to_value(initial_witness, input.witness)?;
        let byte = witness_value.to_be_bytes().last().unwrap();
        result.push(*byte);
```

nit

File: acir/src/circuit/opcodes/black_box_function_call.rs [hunk 4]
```diff
@@ -7,138 +7,376 @@ use serde::{Deserialize, Serialize};
...
+            BlackBoxFunc::EcdsaSecp256k1 => BlackBoxFuncCall::EcdsaSecp256k1 {
+                public_key_x: inputs[0..32].to_vec(),
+                public_key_y: inputs[32..64].to_vec(),
+                signature: inputs[64..128].to_vec(),
+                message: inputs[128..].to_vec(),
+                output: outputs[0],
+            },
+            BlackBoxFunc::FixedBaseScalarMul => {
+                BlackBoxFuncCall::FixedBaseScalarMul { input: inputs[0], outputs }
+            }
```
Comment (TomAFrench) [inline acir/src/circuit/opcodes/black_box_function_call.rs:293]: These will panic rather than returning an error if the ACIR is malformed. Can you change this to return an `InvalidData` error?

File: acir/src/circuit/opcodes/black_box_function_call.rs [hunk 5]
```diff
@@ -7,138 +7,376 @@ use serde::{Deserialize, Serialize};
...
+                inputs
+            }
+            BlackBoxFuncCall::SchnorrVerify {
+                public_key_x,
+                public_key_y,
+                signature,
+                message,
+                ..
+            } => {
+                let mut inputs = Vec::new();
```
Comment (TomAFrench) [inline acir/src/circuit/opcodes/black_box_function_call.rs:189]: We can preallocate the length of this vector (and others in this function.

File: acir/src/circuit/opcodes/black_box_function_call.rs [hunk 6]
```diff
@@ -7,138 +7,376 @@ use serde::{Deserialize, Serialize};
...
+                inputs.push(*public_key_x);
+                inputs.push(*public_key_y);
+                inputs.extend(signature.iter().cloned());
```
Comment (TomAFrench) [inline acir/src/circuit/opcodes/black_box_function_call.rs:192]:
```suggestion
                inputs.extend(signature.iter().copied());
```
Can you change the other instances of `.cloned()` to `.copied()`. This signals that we're doing something cheap here.

File: acvm/src/pwg/signature/ecdsa.rs [hunk 7]
```diff
@@ -3,45 +3,42 @@ use std::collections::BTreeMap;
...
+            panic!("pub_key_x should be 32 bytes long, found {} bytes", public_key_x_inputs.len())
+        });
+    let pub_key_y: [u8; 32] =
+        to_u8_vec(initial_witness, public_key_y_inputs)?.try_into().unwrap_or_else(|_| {
+            panic!("pub_key_y should be 32 bytes long, found {} bytes", public_key_y_inputs.len())
+        });
+    let signature: [u8; 32] =
+        to_u8_vec(initial_witness, signature_inputs)?.try_into().unwrap_or_else(|_| {
+            panic!("signature should be 64 bytes long, found {} bytes", signature_inputs.len())
+        });
```
Comment (TomAFrench) [inline acvm/src/pwg/signature/ecdsa.rs:38]: nice.

File: acvm/src/pwg/signature/ecdsa.rs [hunk 8]
```diff
@@ -3,45 +3,42 @@ use std::collections::BTreeMap;
...
 
     let mut hashed_message = Vec::new();
-    for msg in inputs_iter {
+    for msg in message_inputs.iter() {
```
Comment (TomAFrench) [inline acvm/src/pwg/signature/ecdsa.rs:41]: We can use `to_u8_vec` again here, no?

File: acir/src/circuit/opcodes/black_box_function_call.rs [hunk 9]
```diff
@@ -7,138 +7,376 @@ use serde::{Deserialize, Serialize};
...
+    // should have 128 bits of security
+    HashToField128Security {
+        inputs: Vec<FunctionInput>,
+        output: Witness,
+    },
+    EcdsaSecp256k1 {
+        public_key_x: Vec<FunctionInput>,
+        public_key_y: Vec<FunctionInput>,
+        signature: Vec<FunctionInput>,
+        message: Vec<FunctionInput>,
```
Comment (TomAFrench) [inline acir/src/circuit/opcodes/black_box_function_call.rs:71]:
```suggestion
        hashed_message: Vec<FunctionInput>,
```

We're pretty consistent in specifying this is the hashed message elsewhere so let's ensure we do it here.

**Comment by sirasistant** [comment]
@TomAFrench I pushed a draft PR to the aztec_backend, will do the same for noir next!

File: acvm/src/pwg/signature/ecdsa.rs (hunk 3 above)
Comment (TomAFrench) [reply acvm/src/pwg/signature/ecdsa.rs]: Ah, that's a shame. Nevermind about moving the `to_be_bytes` call then.

File: acvm/src/pwg/signature/ecdsa.rs [hunk 10]
```diff
@@ -3,45 +3,42 @@ use std::collections::BTreeMap;
...
+    initial_witness: &mut BTreeMap<Witness, FieldElement>,
+    public_key_x_inputs: &[FunctionInput],
+    public_key_y_inputs: &[FunctionInput],
+    signature_inputs: &[FunctionInput],
+    message_inputs: &[FunctionInput],
+    output: Witness,
+) -> Result<OpcodeResolution, OpcodeResolutionError> {
+    let pub_key_x: [u8; 32] =
+        to_u8_vec(initial_witness, public_key_x_inputs)?.try_into().unwrap_or_else(|_| {
+            panic!("pub_key_x should be 32 bytes long, found {} bytes", public_key_x_inputs.len())
```
[the remark this answers is not in the document]
Comment (TomAFrench) [reply acvm/src/pwg/signature/ecdsa.rs:29]: We have similar code up in `aztec_backend` which returns a `OpcodeResolutionError::BlackBoxFunctionFailed(func_call.name, !("expected pubkey size {} but received {}", 64, v.len()))

https://github.com/noir-lang/aztec_backend/pull/151/files#diff-88dc9a77ad501e05238d37c3ed80c6d1016dfe0070d7f694db546a0594262b1dR88-R93`

File: acir/src/circuit/opcodes/black_box_function_call.rs [hunk 11]
```diff
@@ -7,138 +7,413 @@ use serde::{Deserialize, Serialize};
...
+    write_u32(&mut writer, num_inputs)?;
+
+    for input in inputs {
+        write_input(input, &mut writer)?;
+    }
+
+    Ok(())
+}
+
+fn write_outputs<W: Write>(outputs: &Vec<Witness>, mut writer: W) -> std::io::Result<()> {
```
Comment (TomAFrench) [inline acir/src/circuit/opcodes/black_box_function_call.rs:101]:
```suggestion
fn write_inputs<W: Write>(inputs: &[FunctionInput], mut writer: W) -> std::io::Result<()> {
    let num_inputs = inputs.len() as u32;
    write_u32(&mut writer, num_inputs)?;

    for input in inputs {
        write_input(input, &mut writer)?;
    }

    Ok(())
}

fn write_outputs<W: Write>(outputs: &[Witness], mut writer: W) -> std::io::Result<()> {
```

**Comment by sirasistant** [comment]
@TomAFrench Pushed a draft PR in the noir repo to use this ACVM version

File: acir/src/circuit/opcodes/black_box_function_call.rs [hunk 12]
```diff
@@ -7,138 +7,413 @@ use serde::{Deserialize, Serialize};
...
+        let mut result = String::new();
+        result += &format!("(_{},...,_{})", first.witness_index(), last.witness_index());
+        result
     }
 }
 
 impl std::fmt::Display for BlackBoxFuncCall {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        let uppercase_name: String = self.name.name().into();
+        let uppercase_name: String = self.get_black_box_func().name().into();
```
Comment (phated) [inline acir/src/circuit/opcodes/black_box_function_call.rs:399]: This seems to be done a lot, why not just have a `.name()` function on `BlackBoxFuncCall`?

File: acvm/src/lib.rs [hunk 13]
```diff
@@ -243,11 +253,22 @@ pub fn default_is_opcode_supported(
...
+        !matches!(opcode, Opcode::BlackBoxFuncCall(BlackBoxFuncCall::AES { .. }) | Opcode::Block(_))
+    }
+
+    match language {
+        Language::R1CS => r1cs_is_supported,
+        Language::PLONKCSat { .. } => plonk_is_supported,
+    }
+}
+
+pub fn default_is_bb_supported(language: Language) -> impl Fn(BlackBoxFunc) -> bool {
```
Comment (TomAFrench) [inline acvm/src/lib.rs:265]: I'd like to avoid adding this function (see deprecation of `default_is_opcode_supported`). At the very least we should immediately deprecate it.

**Review by TomAFrench — COMMENTED** [review]
Thanks for opening up the draft PRs! This is looking good to me. Just a couple of notes here and there and then we can start merging this in.

File: acvm/src/lib.rs (hunk 13 above)
Comment (TomAFrench) [reply acvm/src/lib.rs:265]: https://github.com/noir-lang/acvm/pull/273/files#r1190651498

Comment (sirasistant) [reply acvm/src/lib.rs:265]: Yeah, I don't know how we are going to check what the backend supports during compilation (at that point there is no backend associated yet)... But it's true that we can avoid adding this function by creating a dummy constructor for BB call variants. I'm going to drop this change in favor of that because it should help with the future vision of "specifying supported opcodes" instead of specifically talking about bb calls!

File: acir/src/circuit/opcodes/black_box_function_call.rs [hunk 14]
```diff
@@ -7,138 +7,416 @@ use serde::{Deserialize, Serialize};
...
-            let num_bits = read_u32(&mut reader)?;
-            let input = FunctionInput { witness, num_bits };
-            inputs.push(input)
+        let inputs = read_inputs(&mut reader)?;
+        let outputs = read_outputs(&mut reader)?;
+
+        match name {
+            BlackBoxFunc::AES => Ok(BlackBoxFuncCall::AES { inputs, outputs }),
+            BlackBoxFunc::AND => {
+                if inputs.len() < 2 || outputs.is_empty() {
```
[the remark this answers is not in the document]
Comment (sirasistant) [reply acir/src/circuit/opcodes/black_box_function_call.rs:256]: I didn't want to be too restrictive but it's true that it signals that the serialized bb call either is corrupt or was serialized with a different version of the acvm. I'm going to make it more restrictive

**Review by TomAFrench — APPROVED** [review]
LGTM
